### ISSUE TITLE
Don't assert xact count changes

### ISSUE THREAD

**Issue by mjsir911** [ledger/ledger#2098]
Hey, I've been patching my local ledger install to make some python scripts that I wrote work, basically this line:
```
assert(xact_count == journal->xacts.size());
```

in `src/session.cc` prevents python scripts from `journal.add_xact`ing new transactions into it. Removing this line fixes it.

I'm sure its there for other reasons, but removing it hasn't destroyed everything for me yet!

**Comment by tbm** [comment]
@xnox do you remember how the Python code works and can you comment on this?

@q3aiml Andy, can you comment on this

@jwiegley can you comment on this

**Comment by tbm** [comment]
I guess this would fix #973?

**Comment by jwiegley** [comment]
I'd be fine with removing that assert, if it's causing problems.

**Comment by mjsir911** [comment]
Of note is that this assert is also triggered with this behavior:
```python
>>> from ledger import session
>>> session.read_journal_files()
<ledger.Journal object at 0x7efecd032040>
>>> session.read_journal_files()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Assertion failed in "/var/tmp/portage/app-office/ledger-3.2.1/work/ledger-3.2.1/src/session.cc", line 183:std::size_t ledger::session_t::read_data(const string&): xact_count == journal->xacts.size()
```

**Comment by tbm** [comment]
@mjsir911 can you create a pull request?

**Comment by tbm** [comment]
Oh, I just read the second part of your message. So you're saying removing the `assert` isn't enough because transactions are duplicated?

**Comment by mjsir911** [comment]
> Oh, I just read the second part of your message. So you're saying removing the `assert` isn't enough because transactions are duplicated?

Right, there are cases where the `assert` is actually protecting against wonky behavior. I'll try to spend some time spelunking through to see where I can trigger this behavior, maybe we can protect against it more explicitly? 

What's the expectation with calling `read_journal_files` multiple times?

Sorry for complicating this bug, just want to make sure everything's in order!

**Comment by afh** [comment]
If I'm not mistaken then this issue has been around for a long time (~10 years). I believe fixing this will unlock a lot of Python related functionality for people, especially in combination with #2196.

I'm happy to help getting this fixed, but all my attempts to make progress on my own have failed miserably and I gave up at some point as I understood too little about the inner workings of Ledger, Boost, and C++.

@jwiegley is this something where you could provide some insight on how to solve this?

**Comment by jwiegley** [comment]
Really, I think we'll have to print out all of the transactions that make up both numbers, to find out what's missing in one of the counts. It's meant to be a safety check, to ensure that the two ways of viewing the input data always match. We'll have to find the reason fro the discrepancy; but I don't have any unique insights beyond that.

**Comment by afh** [comment]
Your use-case may be different, yet I've been able to load a ledger journal twice without the assertion triggering by calling `ledger.close_journal_files()` inbetween.

I think there are more likely related bugs in regards to opening, reading, and closing ledger journals in python. I haven't understood the various issues, and possibly it is lack of my understanding of the APIs. I'll come up with better test files to demonstrate the issues or difficulties in my understanding.


```
#!/usr/bin/env PYTHONPATH=build python

import unittest

import ledger

class TestLedgerReadJournal(unittest.TestCase):

    def test_multipl_read_journal(self):
        journal_file = "test/input/drewr3.dat"
        journal = ledger.read_journal(journal_file)
        xacts = len(journal)
        self.assertGreater(xacts, 0)

        ledger.close_journal_files()

        journal = ledger.read_journal(journal_file)
        self.assertEqual(xacts, len(journal))

if __name__ == '__main__':
    unittest.main()
```

**Comment by jwiegley** [comment]
It sounds like folks are being bitten by the fact that the xact count is part of the session state, and isn't being reset when people are expecting it to happen. Indeed, `close_journal_files` should explicitly reset this value.
